lib/paperclip/lazy_thumbnail.rb

```diff
@@ -1,24 +1,128 @@
+      end
+    end
+
+    def initialize(file, options = {}, attachment = nil)
+      super
+
+      @crop = options[:geometry].to_s[-1, 1] == '#'
+      @current_geometry = options.fetch(:file_geometry_parser, Geometry).from_file(@file)
+      @target_geometry = options[:pixels] ? PixelGeometryParser.parse(@current_geometry, options[:pixels]) : options.fetch(:string_geometry_parser, Geometry).parse(options[:geometry].to_s)
+      @format = options[:format]
+      @current_format = File.extname(@file.path)
+      @basename = File.basename(@file.path, @current_format)
+
+      correct_current_format!
+    end
+
     def make
       return File.open(@file.path) unless needs_convert?
 
-      if options[:geometry]
-        min_side = [@current_geometry.width, @current_geometry.height].min.to_i
-        options[:geometry] = "#{min_side}x#{min_side}#" if @target_geometry.square? && min_side < @target_geometry.width
-      elsif options[:pixels]
-        width  = Math.sqrt(options[:pixels] * (@current_geometry.width.to_f / @current_geometry.height)).round.to_i
-        height = Math.sqrt(options[:pixels] * (@current_geometry.height.to_f / @current_geometry.width)).round.to_i
-        options[:geometry] = "#{width}x#{height}>"
+      dst = TempfileFactory.new.generate([@basename, @format ? ".#{@format}" : @current_format].join)
+
+      if preserve_animation?
+        if @target_geometry.nil? || (@current_geometry.width <= @target_geometry.width && @current_geometry.height <= @target_geometry.height)
+          target_width = 'iw'
+          target_height = 'ih'
+        else
+          scale = [@target_geometry.width.to_f / @current_geometry.width, @target_geometry.height.to_f / @target_geometry.height].min
+          target_width = (@current_geometry.width * scale).round
+          target_height = (@current_geometry.height * scale).round
+        end
+
+        # The only situation where we use crop on GIFs is cropping them to a square
+        # aspect ratio, so this is the only special case we implement. If cropping
+        # ever becomes necessary for other situations, this will need to be expanded.
+        if @target_geometry&.square?
+          crop_width = [target_width, target_height].min
+          crop_height = [target_width, target_height].min
+        end
+
+        filter = begin
+          if @crop
+            "scale=#{target_width}:#{target_height}:force_original_aspect_ratio=increase,crop=#{crop_width}:#{crop_height}"
+          else
+            "scale=#{target_width}:#{target_height}:force_original_aspect_ratio=decrease"
+          end
+        end
+
+        command = Terrapin::CommandLine.new(Rails.configuration.x.ffmpeg_binary, '-nostdin -i :source -map_metadata -1 -filter_complex :filter -y :destination', logger: Paperclip.logger)
+        command.run({ source: @file.path, filter: "fps=#{GIF_FPS},#{filter},split[a][b];[a]palettegen=max_colors=#{GIF_PALETTE_COLORS}[p];[b][p]paletteuse=dither=bayer", destination: dst.path })
```

Comment:
Why hardcode a GIF framerate? Shouldn't we use the same exact frames as the source?

Reply:
The idea is to keep the file size down so that more avatars/headers fit under the limit.

Reply:
Maybe we can set a max framerate (24ips?)

Reply:
GIFs can have variable framerate, I think providing a fixed FPS may actually increase the resulting file size if the source file has a lower effective average framerate, but this probably needs some more testing

Reply:
Changed it to a 60 fps max framerate, and a maximum of 3000 frames. Looking at GIF files known to my server, many GIFs are 30, 33 or 50 fps and would feel significantly degraded with a 12 fps hardcoded framerate.

Some GIFs known to my server are 100 fps, but that is most probably a waste we can avoid.